theseus/embodied/motionmodel/misc.py

```diff
@@ -0,0 +1,160 @@
+                raise ValueError("Limit and threshold must be scalar variables.")
+        if self.threshold.tensor.max() < 0.0:
+            raise ValueError("The threshold must be a positive scalar.")
+        self.register_optim_var("vector")
+        self.register_aux_vars(["limit", "threshold"])
+        if side not in ["below", "above", "both"]:
+            raise ValueError("side must be one of 'both', 'above', 'below'.")
+        self.side = side
+        self.dims = dims
+
+    def dim(self):
+        if self.dims is not None:
+            return len(self.dims)
+        return self.vector.dof()
+
+    def _compute_error(self) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        vector = self.vector.tensor
+        limit = self.limit.tensor
+        threshold = self.threshold.tensor
+        down_limit = -limit + threshold
+        up_limit = limit - threshold
+        below_idx = vector < down_limit
+        above_idx = vector > up_limit
+        base_error = vector.new_zeros(vector.shape)
+        if self.side in ["below", "both"]:
+            base_error = torch.where(below_idx, down_limit - vector, base_error)
+        if self.side in ["above", "both"]:
+            base_error = torch.where(above_idx, vector - up_limit, base_error)
+        error = (
+            base_error
+            if self.dims is None
+            else base_error[:, self.dims].view(-1, self.dim())
+        )
+        return error, below_idx, above_idx
+
+    def error(self) -> torch.Tensor:
+        return self._compute_error()[0]
+
+    def jacobians(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        error, below_idx, above_idx = self._compute_error()
+        raw_J = self.vector.tensor.new_zeros(
+            self.vector.shape[0], self.vector.dof(), self.vector.dof()
+        )
+        if self.side in ["below", "both"]:
+            raw_J[below_idx.diag_embed()] = -1.0
+        if self.side in ["above", "both"]:
+            raw_J[above_idx.diag_embed()] = 1.0
+
+        J = (
+            raw_J
+            if self.dims is None
+            else raw_J[:, self.dims, :].view(-1, self.dim(), self.vector.dof())
+        )
+
+        return [J], error
+
+    def _copy_impl(self, new_name: Optional[str] = None) -> "HingeLoss":
+        return HingeLoss(
+            self.vector.copy(),
+            self.limit.copy(),
+            self.threshold.copy(),
+            self.weight.copy(),
+            name=new_name,
+            side=self.side,
+            dims=self.dims,
+        )
+
+
+class Nonholonomic(CostFunction):
+    def __init__(
+        self,
+        pose: Union[SE2, Point3, Vector],
+        vel: Union[Point3, Vector],
+        cost_weight: CostWeight,
+        name: Optional[str] = None,
+    ):
+        super().__init__(cost_weight, name=name)
+        if vel.dof() != 3 or pose.dof() != 3:
+            raise ValueError(
+                "Nonholonomic only accepts 3D velocity or poses (x, y, z dims). "
```

Comment:
x, y, theta dims

Reply:
Fixed in 9a7cb885bfffa9bdb2fd8d41c77bf6563f61e666.